src/main/scala/io/iohk/ethereum/db/storage/CachedKeyValueStorage.scala

```diff
@@ -0,0 +1,39 @@
+package io.iohk.ethereum.db.storage
+
+import io.iohk.ethereum.common.SimpleMap
+import io.iohk.ethereum.db.cache.Cache
+import scala.collection._
+
+trait CachedKeyValueStorage[K, V, T <: CachedKeyValueStorage[K, V, T]] extends SimpleMap[K, V, T]{
+  type I <: KeyValueStorage[K, V, I]
+  protected val storage: I
+  protected val cache: Cache[K, V]
+  def apply(cache: Cache[K,V] , storage: I): T
+
+  def get(key: K): Option[V] = cache.get(key) orElse storage.get(key)
+
+  def update(toRemove: Seq[K], toUpsert: Seq[(K, V)]): T =  {
+    cache.update(toRemove, toUpsert)
+    apply(cache, storage)
+  }
+
+  def updateCond(toRemove: Seq[K], toUpsert: Seq[(K, V)], inMemory: Boolean): T =  {
+    if (inMemory)
+      cache.update(toRemove, toUpsert)
+    else
+      storage.update(toRemove, toUpsert)
+
+    apply(cache, storage)
+  }
+
+  def persist(): Boolean = {
+    if (cache.shouldPersist) {
```

Comment:
Here the cache is dumped to disk and everything is cleared. Have we considered doing something like a _sliding window_ where we only dump old stuff that has survived? I'm just curious about it, I don't think that should be done within this task.

Reply:
I was considering it but taking into account the timeline for 1.1 i came to the conclusion I won't make it in time and it is better to have this simple flush all semantics and implement sliding window maybe for 1.2 .

Reply:
Cool. Should we add a task marked for future releases?